Week 8 Independant Lab: Advanced Data Manipulation <br>

Author: Parker Munsey <br>
Course: BGEN632 Grad. Intro. to Python <br>
Term: Spring 2025 <br>
Date Due: April 21, 2025 <br>

In [2]:
# Import libraries
import pandas as pd

# Load the dataset
df = pd.read_csv(r"C:\Users\TechnellogicPC\UTW\Python\week8labs\gh_turnover_four_projects.csv")

# Rename the specified columns
df.rename(columns={
    'domain': 'project_domain',
    'language': 'project_language',
    'windows': 'project_age',
    'window_idx': 'quarter',
    'num_team': 'team_size',
    'num_commits': 'project_commits',
    'blau_gender': 'gender_blau',
    'Gini_gh_ten': 'gh_tenure_gini',
    'Core1': 'core_dev',
    'commits': 'user_commits',
    'propCommits': 'commit_proportion',
    'github_tenure': 'gh_tenure',
    'leavesNextQ': 'leaves_next_q'
}, inplace=True)

# 1. Calculate the mean of user_commits
mean_user_commits = df['user_commits'].mean()
print("Mean of user_commits (all data):", mean_user_commits)

# 2. Calculate the median of user_commits
median_user_commits = df['user_commits'].median()
print("Median of user_commits (all data):", median_user_commits)

# 3. Select columns that start with 'p' or contain 'g'
turnover_new = df.filter(regex='(^p)|(g)')
print("Columns in turnover_new:")
print(turnover_new.columns.tolist())

# 4. Filter turnover_new where project_commits > 9000
filtered_turnover_new = turnover_new[turnover_new['project_commits'] > 9000]

# 5. Combine steps 3 & 4 with piping
turnover_newer = (
    df
    .filter(regex='(^p)|(g)')
    .loc[lambda d: d['project_commits'] > 9000]
)

# Optional: display a few rows
print("Sample of turnover_newer:")
print(turnover_newer.head())

# 6. From original df, select columns ending with 'e' or containing 's', then filter rows
tenure_filtered_df = (
    df
    .filter(regex='(e$)|s')
    .loc[lambda d: df['gh_tenure'] >= 1991]
)

# 7. Calculate mean and median of user_commits from filtered rows
filtered_mean = df.loc[tenure_filtered_df.index, 'user_commits'].mean()
filtered_median = df.loc[tenure_filtered_df.index, 'user_commits'].median()

print("Mean of user_commits (filtered):", filtered_mean)
print("Median of user_commits (filtered):", filtered_median)


Mean of user_commits (all data): 339.36575052854124
Median of user_commits (all data): 18.0
Columns in turnover_new:
['project_id', 'project_domain', 'project_language', 'project_age', 'project_commits', 'gender_blau', 'gh_tenure_gini', 'gender', 'gh_tenure']
Sample of turnover_newer:
   project_id project_domain project_language  project_age  project_commits  \
0        1101          OTHER           Python           23            22688   
1        1101          OTHER           Python           23            22688   
2        1101          OTHER           Python           23            22688   
3        1101          OTHER           Python           23            22688   
4        1101          OTHER           Python           23            22688   

   gender_blau  gh_tenure_gini gender  gh_tenure  
0          0.0        0.422259   male        426  
1          0.0        0.422259   male         46  
2          0.0        0.422259   male        941  
3          0.0        0.422259   ma

### Comparison of User Commit Activity

In this analysis, we calculated the mean and median of user_commits across the full dataset and a filtered subset:

- **Full Dataset**:
  - Mean: ~339.37
  - Median: 18.0

- **Filtered Dataset** (columns ending in 'e' or containing 's', and `gh_tenure >= 1991`):
  - Mean: 3.5
  - Median: 3.5

This significant drop suggests that contributors with long tenure and matching the filtered criteria have much lower recent contribution counts. This could reflect a shift in involvement or role over time.


References: <br>

Pandas Official Documentation <br>
  https://pandas.pydata.org/docs/   <br>
  Used to reference syntax for `filter()`, `loc[]`, and aggregation functions like `mean()` and `median()`. <br>

Regular Expression Syntax <br>
  https://docs.python.org/3/library/re.html  <br>
  Regex expressions like `(^p)`, `(g)`, and `(e$)` were used to filter columns based on names. These patterns help identify columns starting with, containing, or ending with specific letters. <br> 

Stack Overflow Community  <br>
  https://stackoverflow.com/questions/29550459/select-columns-containing-a-string-in-pandas-dataframe  <br>
  Reviewed examples of how to select columns using regex in Pandas. <br> 